### PR TITLE
feat: Added reverse feature

### PR DESCRIPTION
**Pull request by M-Izadmehr**
There are many cases in which we need inverted/reversed sliders (#216 - #235 - #183 - #86). One important case is for RTL websites. The inverted feature is more general than RTL which can also be used for vertical sliders.

There was a previous pull request trying to add this feature (#334), which had several problems:
- It did not include correct keyboard listeners:
    - in the horizontal inverted slider, the right/left keys should subtract/add
    - in the vertical inverted slider, the up/bottom arrow keys should subtract/add
- The values were not updated correctly:
    - In the inverted component, the value of the right side (top in vertical) is the starting value.
    - The ratio and relative value should be calculated from the width of component (which was subtracted by constant value 100, in that pull request).

As a result, I created this pull request to handle this feature by simply adding an inverted prop:
`<Slider reverse />`

### CODE REVIEW

**Comment by M-Izadmehr** [comment]
Yeah, that is a good suggestion, I will fix it asap:)

**Comment by tonymaibox** [comment]
Oooo, this is v. nice! Would love to see this merged and released soon 🙏 

Thanks for this PR.

File: README.md [hunk 1]
```diff
@@ -101,6 +101,7 @@ The following APIs are shared by Slider and Range.
 | vertical | boolean | `false` | If vertical is `true`, the slider will be vertical. |
 | handle | (props) => React.ReactNode | | A handle generator which could be used to customized handle. |
 | included | boolean | `true` | If the value is `true`, it means a continuous value interval, otherwise, it is a independent value. |
+| inverted | boolean | `false` | If the value is `true`, it means the component is rendered inverted. |
```
Comment (zombieJ) [inline README.md]: here : )

Comment (M-Izadmehr) [reply README.md]: Done :))

**Comment by M-Izadmehr** [comment]
> Hi @M-Izadmehr , I clone the code and find that `slider.html` sample has some click interactive issue. Could you pls take a look?
> kapture 2018-10-21 at 22 24 21
Sorry for my delayed response, I was quite busy in the past few weeks, and I did not have the time to work on this issue.
This problem raised from the fact that the margin-left of the handles was handled in the styles section. I used the transform property to dynamically move the center of handle to the end-point of the track. Now this problem is fixed in addition to another issue in reverse marks of vertical sliders.

**Comment by zombieJ** [comment]
Hi @M-Izadmehr , Current transform breaks the origin example:

**Comment by M-Izadmehr** [comment]
> Hi @M-Izadmehr , Current transform breaks the origin example:
> 

You are right on that, although I believe transform offers us a more general solution, it breaks the backward compatibility. As a result, I reverted the handle origin changes to the way it was.

**Comment by barakplasma** [comment]
@zombieJ I also took a look at this PR, and thanks to @M-Izadmehr 's changes, I couldn't recreate your screenshot.
m-izadmehr-master-ad062e817436e00b603b2f2e83e680791dbc06e4

I am excited by the prospect of using M-Izadmehr's changes, is there anything I can do to help merge this PR?

**Comment by barakplasma** [comment]
@M-Izadmehr I forked your repo and added some examples in https://github.com/M-Izadmehr/slider/pull/1/commits/c139d1b51c35a5dc0466fb73402f82d376f942e7 .
I did a tiny bit of QA and noticed that the marks end label is a bit off at https://barakplasma.github.io/slider/examples/marks.html in the "Reversed Slider with marks and steps" example I added.
screen shot 2019-02-20 at 10 55 55
I might be able to look into this if you are interested.

P.S. if others want to see the reversed rc-slider, I host them at https://barakplasma.github.io/slider/

**Comment by barakplasma** [comment]
@M-Izadmehr @zombieJ I took the initiative to fix the markLabels being far away from the marks in my PR https://github.com/M-Izadmehr/slider/pull/1/commits/95d300cb5896ab0bdc93cb2f1e9c39edb40927d5 
screen shot 2019-03-04 at 16 54 58

Please take a look at my PR, and let me know what you think. (alternatively, check out my gh-pages version at https://barakplasma.github.io/slider/examples/marks.html )

**Comment by barakplasma** [comment]
Another thing: 
I think in @types/rc-slider 's index.d.ts `export interface SliderProps extends CommonApiProps {` will need to be updated on line 143 to have the following:
```typescript
    /**
     * Reverse the direction of the slider.
     * From Left to Right To Right to Left
     * @default false
     */
    reverse?: boolean;
```

**Comment by zombieJ** [comment]
Thanks @barakplasma ,
@M-Izadmehr could you help to check on this?

**Comment by Joysome** [comment]
Any chances of having this soon?

**Comment by M-Izadmehr** [comment]
@barakplasma Thank you for the pull request. 
Sorry, I was not active for a while. Now, 
[1] I merged pull requests from @barakplasma 
[2] Created a new [https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35653](pull request)  for typescript interface

I think this pull request is in a good state, to be merged now.

**Comment by theinen** [comment]
Hey, just following up to see if this can be merged soon. Fixes a big issue for me :)

**Comment by zombieJ** [comment]
Seems still not align with center:
Kapture 2019-06-17 at 12 20 36

Let me check...

**Comment by barakplasma** [comment]
I don't understand, what's not aligned with the center of what?

**Comment by zombieJ** [comment]
Handle not center align with the mark point.

**Comment by M-Izadmehr** [comment]
> Handle not center align with the mark point.

I just checked this issue, this is exactly the behaviour that we expect:

1) in normal slider (without `marks`) we would like the slider to move with the mouse.
2) when there are marks, we want the slider to only stay on marks, and not in between.
3) when we have steps, we only want the slider to stay on steps, and not in between.

In the main [demo](http://react-component.github.io/slider/examples/marks.html) page you can check this behaviour:
ezgif-4-93dd5e5d1939

Which is similar to the reverse behaviour:
reverse

**Comment by zombieJ** [comment]
Not mean the pin effect. I mean the handle point (big circle) center is not align the center of mark point (small circle)

**Comment by M-Izadmehr** [comment]
Yeah, thank you very much for mentioning that.
I fixed it by adding a different style for reverse dots.
Now the match perfectly on each other.
image

**Comment by hadiranjbar25** [comment]
The pull request is still open. Is there any plan to publish this feature?

**Comment by M-Izadmehr** [comment]
I think it is now in a good stat to be published:)

**Pull request merged**